connectors/src/connectors/intercom/temporal/sync_help_center.ts

```diff
@@ -0,0 +1,383 @@
+import type { ModelId } from "@dust-tt/types";
+import type { Client as IntercomClient } from "intercom-client";
+
+import type { IntercomCollectionType } from "@connectors/connectors/intercom/lib/intercom_api";
+import {
+  fetchIntercomArticles,
+  fetchIntercomCollection,
+  fetchIntercomCollections,
+  fetchIntercomHelpCenter,
+} from "@connectors/connectors/intercom/lib/intercom_api";
+import {
+  getHelpCenterArticleDocumentId,
+  getHelpCenterCollectionDocumentId,
+} from "@connectors/connectors/intercom/lib/utils";
+import {
+  deleteFromDataSource,
+  renderDocumentTitleAndContent,
+  upsertToDatasource,
+} from "@connectors/lib/data_sources";
+import type { Connector } from "@connectors/lib/models";
+import type { IntercomHelpCenter } from "@connectors/lib/models/intercom";
+import {
+  IntercomArticle,
+  IntercomCollection,
+} from "@connectors/lib/models/intercom";
+import logger from "@connectors/logger/logger";
+import type { DataSourceConfig } from "@connectors/types/data_source_config";
+
+/**
+ * This function syncs a Help Center and its children (collections & articles)
+ * from Intercom to the database and the core data source.
+ *
+ * If the Help Center is not on Intercom anymore or if we have revoked permissions:
+ * -> We delete it and its children from the database and the core data source.
+ *
+ * If the Help Center is on Intercom and we have read permissions:
+ * -> We loop on all level 1 collections of the Help Center and sync them.
+ */
+export async function syncHelpCenter({
+  connector,
+  intercomClient,
+  dataSourceConfig,
+  helpCenter,
+  loggerArgs,
+}: {
+  connector: Connector;
+  intercomClient: IntercomClient;
+  dataSourceConfig: DataSourceConfig;
+  helpCenter: IntercomHelpCenter;
+  loggerArgs: Record<string, string | number>;
+}) {
+  const connectorId = connector.id;
+  const helpCenterOnIntercom = await fetchIntercomHelpCenter(
+    connector.connectionId,
+    helpCenter.helpCenterId
+  );
+
+  // If our rights were revoked or the help center is not on intercom anymore we delete it
+  if (!helpCenterOnIntercom || helpCenter.permission === "none") {
+    const level1Collections = await IntercomCollection.findAll({
+      where: {
+        connectorId: connector.id,
+        helpCenterId: helpCenter.helpCenterId,
+        parentId: null,
+      },
+    });
+    await Promise.all(
+      level1Collections.map(async (collection) => {
+        await _deleteCollection({
+          connectorId,
+          dataSourceConfig,
+          collection,
+          loggerArgs,
+        });
+      })
+    );
+    await helpCenter.destroy();
+    return;
+  }
+
+  // Otherwise we update its name and sync its collections
+  await helpCenter.update({ name: helpCenterOnIntercom.display_name });
+  const level1Collections = await IntercomCollection.findAll({
+    where: {
+      connectorId,
+      helpCenterId: helpCenter.helpCenterId,
+      parentId: null,
+    },
+  });
+  await Promise.all(
```

Comment:
We could push further and after potentially deleting what needs to be deleted, return the set of collections to sync so that we call syncCollection in separate activities.

If these collection are big we'll retry everything at each attempt?

As mentioned in a comment below maybe we can be smarter about not retrying everything if we are in a retry attempt of that activity?

Again if the cardinal of articles can be big under that function we should probably return all the ids we need to pull and upsert and then process these chunk by chunk to limit the blast of a failed acitvity?

Reply:
OK I think it would be nice to not retry upserting everything if we just upserted it when we fail and retry an activity. This should strike a nice 80/20 here (just a matter of passing around a workflow start timestamp and act on it)?

Reply:
I split in multiple activities because it's easy to do: 
-> activity to update the help center object, and remove everything from db if we're not allowed to access it anymore.
-> activity to fetch the list of collections to upsert / remove
-> activity to process a collection.